runtime/lua/vim/diagnostic.lua

```diff
@@ -755,7 +755,31 @@ function M.get(bufnr, opts)
     opts = { opts, 't', true },
   })
 
-  return get_diagnostics(bufnr, opts, false)
+  return vim.deepcopy(get_diagnostics(bufnr, opts, false))
+end
+
+--- Get current diagnostics count.
+---
+---@param bufnr integer|nil Buffer number to get diagnostics from. Use 0 for
+---                        current buffer or nil for all buffers.
+---@param opts table|nil A table with the following keys:
+---                        - namespace: (number) Limit diagnostics to the given namespace.
+---                        - lnum: (number) Limit diagnostics to the given line number.
+---                        - severity: See |diagnostic-severity|.
+---@return table A table with actually present severity values as keys (see |diagnostic-severity|) and integer counts as values.
+function M.get_count(bufnr, opts)
+  vim.validate({
+    bufnr = { bufnr, 'n', true },
+    opts = { opts, 't', true },
+  })
+
+  -- A faster version of `get()` which doesn't create copies
+  local diagnostics = get_diagnostics(bufnr, opts, false)
+  local count = {}
+  for _, d in ipairs(diagnostics) do
+    count[d.severity] = (count[d.severity] or 0) + 1
+  end
```

Comment:
Since the _raison d'être_ of this function is performance, might as well go all-in:

```suggestion
  for i = 1, #diagnostics do
    local d = diagnostics[i]
    count[d.severity] = (count[d.severity] or 0) + 1
  end
```

Reply:
I see this optimization and raise with `local severity = diagnostics[i].severity`.